### PR TITLE
Add support for encrypting secondary user store password with symmetric encryption

### PR DESCRIPTION
**Pull request by denuwanthi**
Resolves https://github.com/wso2/product-is/issues/7905
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java [hunk 1]
```diff
@@ -144,8 +135,21 @@ private void initializeKeyStore() throws IdentityUserStoreMgtException {
...
+            String encryptionKeyStore = config.getFirstProperty(ENCRYPTION_KEYSTORE);
+            if (INTERNAL_KEYSTORE.equalsIgnoreCase(encryptionKeyStore)) {
+                try {
+                    return CryptoUtil.getDefaultCryptoUtil().encryptAndBase64Encode(plainText.getBytes());
+                } catch (CryptoException e) {
+                    String errorMessage = "Error while encrypting the plain text using internal keystore.";
+                    throw new IdentityUserStoreMgtException(errorMessage, e);
+                }
+            }
+        }
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: So how would this switch to symmetric encryption

Comment (malithie) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: This should be fixed like this right.
- Precedence should be given to the 'UserStorePasswordEncryption' config
- If not present, should switch to default internal encryption mechanism

Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: > So how would this switch to symmetric encryption

Switching to symmetric is done on crypto service layer from line 144

Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: > This should be fixed like this right.
> 
> * Precedence should be given to the 'UserStorePasswordEncryption' config
> * If not present, should switch to default internal encryption mechanism
Yes the steps should be like that.
Here 'ENCRYPTION_KEYSTORE' constants is the UserStorePasswordEncryption config.

File: components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java [hunk 2]
```diff
@@ -118,6 +121,40 @@ public static Cipher getCipherOfSuperTenant() throws UserStoreException {
...
+            }
+            String cryptoProvider = config.getFirstProperty(UserStoreConfigurationConstants.CRYPTO_PROVIDER);
+            if (UserStoreConfigurationConstants.SYMMETRIC_KEY_CRYPTO_PROVIDER.equalsIgnoreCase(cryptoProvider)) {
+                isSymmetricKeyEncryptionEnabled = true;
+            }
+        }
+
+        if (isInternalKeyStoreEncryptionEnabled && isSymmetricKeyEncryptionEnabled) {
+
+            throw new CryptoException(String.format("Userstore encryption cazn not be supported due to " +
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: typo 'cazn'

File: components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java [hunk 3]
```diff
@@ -118,6 +121,40 @@ public static Cipher getCipherOfSuperTenant() throws UserStoreException {
...
+                            "conflicting configurations: '%s' and '%s'. When using internal keystore, assymetric crypto " +
+                            "provider should be used.", UserStoreConfigurationConstants.INTERNAL_KEYSTORE,
+                    UserStoreConfigurationConstants.SYMMETRIC_KEY_CRYPTO_PROVIDER));
+        } else if (isInternalKeyStoreEncryptionEnabled || isSymmetricKeyEncryptionEnabled) {
+
+            return CryptoUtil.getDefaultCryptoUtil().encrypt(plainTextBytes);
+
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: Let's remove additional new line

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java [hunk 4]
```diff
@@ -144,11 +138,48 @@ private void initializeKeyStore() throws IdentityUserStoreMgtException {
...
+            }
+            String cryptoProvider = config.getFirstProperty(CRYPTO_PROVIDER);
+            if (SYMMETRIC_KEY_CRYPTO_PROVIDER.equalsIgnoreCase(cryptoProvider)) {
+                isSymmetricKeyEncryptionEnabled = true;
+            }
         }
 
+        if (isInternalKeyStoreEncryptionEnabled && isSymmetricKeyEncryptionEnabled) {
+
+            throw new IdentityUserStoreMgtException(String.format("Userstore encryption cazn not be supported due to " +
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: typo

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java [hunk 5]
```diff
@@ -144,11 +138,48 @@ private void initializeKeyStore() throws IdentityUserStoreMgtException {
...
+            try {
+                return CryptoUtil.getDefaultCryptoUtil().encryptAndBase64Encode(plainText.getBytes());
+            } catch (CryptoException e) {
+                String errorMessage = "Error while encrypting the plain text using internal keystore.";
+                throw new IdentityUserStoreMgtException(errorMessage, e);
+            }
+        } else {
+            return encryptWithPrimaryKeyStore(config, plainText);
+        }
+
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: unnecessary line

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java [hunk 6]
```diff
@@ -144,11 +138,48 @@ private void initializeKeyStore() throws IdentityUserStoreMgtException {
...
+    }
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: What's the difference of this and the one in UserCoreUtil.
Isn't same logic is duplicated in both places

Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: Those 2 classes are in 2 different OSGI bundles.
UserStoreUtil - org.wso2.carbon.identity.user.store.configuration.deployer
SecondaryUserStoreConfigurator. - org.wso2.carbon.identity.user.store.configuration

File: components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java (hunk 2 above)
Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: Fixed with 745688e61d9965c6f573613694fe2fa661f96ac1

File: components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java (hunk 3 above)
Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: Fixed with 745688e61d9965c6f573613694fe2fa661f96ac1

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java (hunk 4 above)
Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: Fixed with 745688e61d9965c6f573613694fe2fa661f96ac1

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java (hunk 5 above)
Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: Fixed with 745688e61d9965c6f573613694fe2fa661f96ac1

File: components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java [hunk 7]
```diff
@@ -118,6 +121,39 @@ public static Cipher getCipherOfSuperTenant() throws UserStoreException {
...
+                            "conflicting configurations: '%s' and '%s'. When using internal keystore, assymetric crypto " +
+                            "provider should be used.", UserStoreConfigurationConstants.INTERNAL_KEYSTORE,
+                    UserStoreConfigurationConstants.SYMMETRIC_KEY_CRYPTO_PROVIDER));
+        } else if (isInternalKeyStoreEncryptionEnabled || isSymmetricKeyEncryptionEnabled) {
+
+            return CryptoUtil.getDefaultCryptoUtil().encrypt(plainTextBytes);
+        } else {
+            return encryptWithPrimaryKeyStore(plainTextBytes);
+        }
+
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: unnecessary line

File: components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java [hunk 8]
```diff
@@ -118,6 +121,39 @@ public static Cipher getCipherOfSuperTenant() throws UserStoreException {
...
+        if (isInternalKeyStoreEncryptionEnabled && isSymmetricKeyEncryptionEnabled) {
+
+            throw new CryptoException(String.format("Userstore encryption can not be supported due to " +
+                            "conflicting configurations: '%s' and '%s'. When using internal keystore, assymetric crypto " +
+                            "provider should be used.", UserStoreConfigurationConstants.INTERNAL_KEYSTORE,
+                    UserStoreConfigurationConstants.SYMMETRIC_KEY_CRYPTO_PROVIDER));
+        } else if (isInternalKeyStoreEncryptionEnabled || isSymmetricKeyEncryptionEnabled) {
+
```
Comment (malithie) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: lets remove this line as well

Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: Fixed with 57ad2c97fc9486e4803a66185de8851642574848

File: components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java (hunk 7 above)
Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration.deployer/src/main/java/org/wso2/carbon/identity/user/store/configuration/deployer/util/UserStoreUtil.java]: Fixed with 57ad2c97fc9486e4803a66185de8851642574848

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java (hunk 6 above)
Comment (denuwanthi) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/SecondaryUserStoreConfigurator.java]: > What's the difference of this and the one in UserCoreUtil.
> Isn't same logic is duplicated in both places

Created a new git issue https://github.com/wso2/product-is/issues/7954 to track this

**Pull request merged**